[mlir][memref] Fix out-of-bounds crash when reifying result dims

matthias-springer opened the pull request:
Do not crash when the input IR is invalid, i.e., when the index of the dimension operand of a `tensor.dim`/`memref.dim` is out-of-bounds. This fixes #70180.

nicolasvasilache commented on mlir/lib/Dialect/MemRef/Transforms/ResolveShapedTypeResultDims.cpp:
```diff
@@ -94,6 +94,9 @@ struct DimOfReifyRankedShapedTypeOpInterface : public OpRewritePattern<OpTy> {
     unsigned resultNumber = dimValue.getResultNumber();
+    // Do not apply pattern if the IR is invalid (dim out of bounds).
+    if (*dimIndex >= reifiedResultShapes[resultNumber].size())
+      return failure();
```
can we systematically add a notifyMatchFailure for a better debugging experience of why patterns fail to apply?

matthias-springer committed:
[mlir][memref] Fix out-of-bounds crash when reifying result dims

Do not crash when the input IR is invalid, i.e., when the index of the dimension operand of a `tensor.dim`/`memref.dim` is out-of-bounds. This fixes #70180.
